File: Trees/865. Smallest Subtree with all the Deepest Nodes/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root: return [None, 0] # [LCA, depth]
            
            leftNode, leftDepth = dfs(root.left)
            rightNode, rightDepth = dfs(root.right)
            
            # smallest lowest common ancestor is root with equal depth of left subtree and right subtree
            if leftDepth == rightDepth:
                return [root, leftDepth+1]
            
            # Lowest Common Ancestor node exists in the subtree with deeper depth
            # don't forget to  update current root's depth. current depth = max(left depth, right depth) + 1
            if leftDepth > rightDepth:
                return [leftNode, leftDepth+1]
            elif leftDepth < rightDepth:
                return [rightNode, rightDepth+1]
                
        return dfs(root)[0]

File: Trees/865. Smallest Subtree with all the Deepest Nodes/solution.cpp
#include <string>
#include <vector>
#include <numeric>
#include <functional>
#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *subtreeWithAllDeepest(TreeNode *root)
    {

        function<pair<TreeNode *, int>(TreeNode *)> dfs = [&](TreeNode *node)
        {
            if (!node)
                return make_pair((TreeNode *)nullptr, 0);

            auto left = dfs(node->left);
            auto right = dfs(node->right);

            if (left.second == right.second)
            {
                // Current node is the common ancestor for deepest nodes in both sides
                return make_pair(node, left.second + 1);
            }

            if (left.second > right.second)
            {
                // Deepest nodes are only in the left
                return make_pair(left.first, left.second + 1);
            }
            else
            {
                // Deepest nodes are only in the right
                return make_pair(right.first, right.second + 1);
            }
        };

        return dfs(root).first;
    }
};